src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/IteratorAndAsyncCaptureWalker.cs

```diff
@@ -204,6 +212,14 @@ private void CaptureVariable(Symbol variable, SyntaxNode syntax)
                 if (_variablesToHoist.Add(variable) && variable is LocalSymbol local && _boundRefLocalInitializers.TryGetValue(local, out var variableInitializer))
                     CaptureRefInitializer(variableInitializer, syntax);
             }
+
+            static bool canRefLocalBeHoisted(LocalSymbol refLocal)
+            {
+                return refLocal.SynthesizedKind == SynthesizedLocalKind.Spill ||
+                    (refLocal.SynthesizedKind == SynthesizedLocalKind.ForEachArray &&
```

Comment:
>refLocal.SynthesizedKind == SynthesizedLocalKind.ForEachArray && [](http://example.com/codeflow?start=21&length=64)

It is not obvious to me that this hoisting is always safe under new conditions. Quite possibly the fact that ref locals were illegal before was sufficient to ensure the safety. Add some suggestions for new test scenarios.